### PR TITLE
FEATURE: control web crawlers access with white/blacklist

### PR DESCRIPTION
**Pull request by nlalonde**
Something we may need to change since the whitelist can block all API requests if they don't include a whitelisted user agent: Currently we include requests without browser-like user agents as web crawlers, including "DiscourseAPI".

### CODE REVIEW

File: lib/middleware/request_tracker.rb [hunk 1]
```diff
@@ -155,12 +157,20 @@ def log_request_info(env, result, info)
...
 
-    if rate_limit(env)
+    if rate_limit(request)
       result = [429, {}, ["Slow down, too Many Requests from this IP Address"]]
       return result
     end
 
+    if block_crawler(request)
+      log_request = false
+      result = [403, {}, []]
```
Comment (SamSaffron) [inline lib/middleware/request_tracker.rb]: we probably want some sort of message explaining this here

File: lib/middleware/request_tracker.rb [hunk 2]
```diff
@@ -266,6 +276,14 @@ def rate_limit(env)
     end
   end
 
+  def block_crawler(request)
+    request.get? &&
+      !request.xhr? &&
```
Comment (SamSaffron) [inline lib/middleware/request_tracker.rb]: I wonder if we should be a bit more atomic in our crawler blocking, but we can defer on this for now and look at numbers first

File: app/models/web_crawler_request.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      write_cache!(Time.now.utc)
+      write_cache!(Time.now.utc.yesterday)
+      return
+    end
+
+    self.last_flush = Time.now.utc
+
+    date = date.to_date
+
+    $redis.smembers(user_agent_list_key(date)).each do |user_agent, _|
```
Comment (SamSaffron) [inline app/models/web_crawler_request.rb]: hmmm how, big is this set getting? should we not be removing all committed rows as we go from the set in an single eval?

Comment (nlalonde) [reply app/models/web_crawler_request.rb]: Yeah we can remove the elements from the set as we write with `spop`. I'll give that a try.

File: lib/middleware/request_tracker.rb (hunk 1 above)
Comment (nlalonde) [reply lib/middleware/request_tracker.rb]: 👍

File: app/assets/javascripts/admin/models/report.js.es6 [hunk 4]
```diff
@@ -114,21 +120,31 @@ const Report = Discourse.Model.extend({
...
+    return this.get('xaxisIsDate') ? data.toArray().reverse() : data.toArray();
+  },
 
-  dataReversed: function() {
-    return this.get("data").toArray().reverse();
-  }.property("data")
+  @computed('data')
+  xaxisIsDate() {
+    if (!this.data[0]) return false;
+    return this.data && moment(this.data[0].x, 'YYYY-MM-DD').isValid();
```
Comment (nlalonde) [inline app/assets/javascripts/admin/models/report.js.es6]: `moment('Wotbox/2.01 (+http://www.wotbox.com/bot/)', 'YYYY-MM-DD').isValid() === true`  :cry:

**Pull request merged**